### PR TITLE
[Vendored Opam] fix partial-match warnings by removing a lot of dead code

### LINKED ISSUE #10216: Compiling dune outputs partial-match warnings
Using the current main branch (at https://github.com/ocaml/dune/commit/f04be44adcf8a42f870d2bf99e8cba17b884bb00), compiling dune currently outputs a slew of warnings:
```
$ ocaml boot/bootstrap.ml
ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
./.duneboot.exe
Done: 445/1525 (jobs: 3)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads stdune.ml
File "otherlibs/stdune/src/stdune.ml", line 1, characters 4-9:
Warning 53 [misplaced-attribute]: the "alert" attribute cannot appear in this context
Done: 454/1525 (jobs: 10)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads fiber.mli
File "vendor/fiber/src/fiber.mli", line 117, characters 3-9:
Warning 53 [misplaced-attribute]: the "inline" attribute cannot appear in this context
Done: 524/1525 (jobs: 10)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads memo.mli
File "src/memo/memo.mli", line 104, characters 3-9:
Warning 53 [misplaced-attribute]: the "inline" attribute cannot appear in this context
Done: 613/1525 (jobs: 10)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads opamLocal.ml
File "vendor/opam/src/repository/opamLocal.ml", lines 109-114, characters 46-34:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _

File "vendor/opam/src/repository/opamLocal.ml", lines 170-189, characters 44-56:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _

File "vendor/opam/src/repository/opamLocal.ml", lines 209-231, characters 8-44:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _
Done: 623/1525 (jobs: 10)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads opamVCS.ml
File "vendor/opam/src/repository/opamVCS.ml", lines 179-182, characters 11-40:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _
Done: 834/1525 (jobs: 8)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads opamDownload.ml
File "vendor/opam/src/repository/opamDownload.ml", lines 323-361, characters 42-41:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _
Done: 851/1525 (jobs: 5)cd _boot && /home/kit_ty_kate/.opam/trunk/bin/ocamlopt.opt -c -g -no-alias-deps -w -49-6 -alert -unstable -I +unix -I +threads opamRepository.ml
File "vendor/opam/src/repository/opamRepository.ml", lines 88-90, characters 45-77:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _

File "vendor/opam/src/repository/opamRepository.ml", lines 277-280, characters 39-59:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Checksum_mismatch _

File "vendor/opam/src/repository/opamRepository.ml", lines 423-427, characters 12-42:
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
(_, Checksum_mismatch _)
```
In particular the `partial-match` warnings can be problematic for some setup that use `OCAMLPARAM=warn-error=+8,_` or similar to check the relative health of the ecosystem (e.g. https://github.com/kit-ty-kate/opam-health-check-ng)

### PR DESCRIPTION
**Pull request by ElectreAAS**
Fixes #10216.
The only functions in `OpamRepository.ml` that are used externally are `pull_file` and `pull_tree`. I removed everything else, as we now have our own fetching code in dune_pkg

### CODE REVIEW

File: vendor/opam/src/repository/opamRepository.ml [hunk 1]
```diff
@@ -115,9 +113,9 @@ let fetch_from_cache =
         | root_cache_url::other_caches ->
           OpamProcess.Job.catch
             (function Failure _
-                    | OpamDownload.Download_fail _ ->
-                      try_cache_dl other_caches
-                    | e -> raise e)
+              | OpamDownload.Download_fail _
+              | OpamDownload.Checksum_mismatch _ -> try_cache_dl other_caches
```
Comment (ElectreAAS) [inline vendor/opam/src/repository/opamRepository.ml:117]: Not sure about this one. This exception could also fall into the re-raise below

**Comment by Leonidas-from-XIV** [comment]
I am honestly not sure of this. I agree that these warnings are annoying but I am also not fond of modifying the vendored code, since it makes it harder and harder to resync with OPAM upstream, especially if they're not part of our [vendored OPAM repo](https://github.com/ocaml-dune/opam) (which is probably also not up to date with current dune).

I think a better way would be to extend the bootstrap process to understand vendored dependencies and make sure they are treated in a (more) similar way to how they are treated in the standard dune build, i.e. do not run with our extended set of warnings.

**Comment by emillon** [comment]
> I agree that these warnings are annoying but I am also not fond of modifying the vendored code, since it makes it harder and harder to resync with OPAM upstream, especially if they're not part of our [vendored OPAM repo](https://github.com/ocaml-dune/opam) (which is probably also not up to date with current dune).

For context, we added these warnings by modifying the vendored code. So while I agree with you that we'd need a more principled way to make changes to our vendored opam code, ignoring the warnings is not the solution here.

**Comment by emillon** [comment]
(in other words, our aim should be decreasing the diff with upstream, not removing code; but in that case that might be aligned)

**Comment by rgrinberg** [comment]
The code for downloading/extracting diffs isn't going to be shared, so in this case, removal should work. Is there a reason why the entire modules can't be just removed? Or replaced with just stubs.

**Comment by kit-ty-kate** [comment]
Hi, this PR is important for opam-health-check-ng to work properly and generally pretty annoying to see when compiling opam from scratch (where dune is vendored), can someone take over this PR and put it over the line?

**Comment by kit-ty-kate** [comment]
If nobody has time to do it i can try to do it, but reading the discussion above, I'm not sure to see what the consensus is on what to do.

**Comment by rgrinberg** [comment]
You can either just revert all of our changes to opam related to this feature or continue along with this PR and remove the code altogether. Both are fine with me. I lean towards removal as there's no plans to share this code anymore and it will reduce the build times for us.

**Pull request closed without merging**